lib/Dialect/FIRRTL/Transforms/InferDomains.cpp

```diff
@@ -0,0 +1,1426 @@
+// Helpers for working with module or instance domain info.
+//====--------------------------------------------------------------------------
+
+/// From a domain info attribute, get the domain-type of a domain value at
+/// index i.
+static StringAttr getDomainPortTypeName(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return nullptr;
+  auto ref = cast<FlatSymbolRefAttr>(info[i]);
+  return ref.getAttr();
+}
+
+/// From a domain info attribute, get the row of associated domains for a
+/// hardware value at index i.
+static auto getPortDomainAssociation(ArrayAttr info, size_t i) {
+  if (info.empty())
+    return info.getAsRange<IntegerAttr>();
+  return cast<ArrayAttr>(info[i]).getAsRange<IntegerAttr>();
+}
+
+/// Return true if the value is a port on the module.
+static bool isPort(FModuleOp module, BlockArgument arg) {
+  return arg.getOwner()->getParentOp() == module;
+}
+
+/// Return true if the value is a port on the module.
+static bool isPort(FModuleOp module, Value value) {
+  auto arg = dyn_cast<BlockArgument>(value);
+  if (!arg)
+    return false;
+  return isPort(module, arg);
+}
+
+//====--------------------------------------------------------------------------
+// Circuit-wide state.
+//====--------------------------------------------------------------------------
+
+/// Each declared domain in the circuit is assigned an index, based on the order
+/// in which it appears. Domain associations for hardware values are represented
+/// as a list of domains, sorted by the index of the domain type.
+using DomainTypeID = size_t;
+
+/// Information about the domains in the circuit. Able to map domains to their
+/// type ID, which in this pass is the canonical way to reference the type
+/// of a domain.
+namespace {
+struct CircuitDomainInfo {
+  CircuitDomainInfo(CircuitOp circuit) { processCircuit(circuit); }
+
+  ArrayRef<DomainOp> getDomains() const { return domainTable; }
+  size_t getNumDomains() const { return domainTable.size(); }
+  DomainOp getDomain(DomainTypeID id) const { return domainTable[id]; }
+
+  DomainTypeID getDomainTypeID(DomainOp op) const {
+    return typeIDTable.at(op.getNameAttr());
+  }
+
+  DomainTypeID getDomainTypeID(StringAttr name) const {
+    return typeIDTable.at(name);
+  }
+
+  DomainTypeID getDomainTypeID(FlatSymbolRefAttr ref) const {
+    return getDomainTypeID(ref.getAttr());
+  }
+
+  DomainTypeID getDomainTypeID(ArrayAttr info, size_t i) const {
+    auto name = getDomainPortTypeName(info, i);
+    return getDomainTypeID(name);
+  }
+
+  DomainTypeID getDomainTypeID(Value value) const {
+    assert(isa<DomainType>(value.getType()));
+    if (auto arg = dyn_cast<BlockArgument>(value)) {
+      auto *block = arg.getOwner();
+      auto *owner = block->getParentOp();
+      auto module = cast<FModuleOp>(owner);
+      auto info = module.getDomainInfoAttr();
+      auto i = arg.getArgNumber();
+      return getDomainTypeID(info, i);
```

Comment:
Consider keeping the logical temporaries for `info` and `i` (and possibly `owner`) and inlining `block` and `module`. Long chains of temporaries that are only there to indicate "the next step" tend to hurt reading comprehension:

```suggestion
      auto *owner = arg.getOwner()->getParentOp();
      auto info = cast<FModuleOp>(owner).getDomainInfoAttr();
      auto i = arg.getArgNumber();
      return getDomainTypeID(info, i);
```

I do recognize that this is a fine balance between "inline everything" (which probably isn't the best here) and "inline enough so that a reader can follow it".

Ditto for the later instance case.

Mega-nit: Most syntax highlighters aren't smart enough to understand when `module` is a keyword and when it isn't. Consider avoiding this and using `moduleOp` when it comes up.